File: medium/cpp/1546.cpp
// https://leetcode.com/problems/maximum-number-of-non-overlapping-subarrays-with-sum-equals-target/

class Solution
{
public:
    int maxNonOverlapping(std::vector<int>& nums, int target)
    {
        const auto n = std::size(nums);

        std::vector<int> dp(n);

        std::unordered_map<int, int> pref;
        pref[0] = -1;

        for (auto i = 0, s = 0; i < n; i++)
        {
            s += nums[i];

            if (i > 0)
            {
                dp[i] = dp[i - 1];
            }

            if (const auto it = pref.find(s - target); it != std::end(pref))
            {
                const auto pos = it->second;

                dp[i] = std::max(dp[i], 1 + (pos == -1 ? 0 : dp[pos]));
            }

            pref[s] = i;
        }

        return dp.back();
    }
};

File: medium/python/1546.py
# 1546. Maximum Number of Non-Overlapping Subarrays With Sum Equals Target

class Solution:
    def maxNonOverlapping(self, nums: List[int], target: int) -> int:
        dp = [0] * len(nums)
        pref = { 0: -1 }
        s = 0

        for i in range(0, len(nums)):
            s += nums[i]

            if i > 0:
                dp[i] += dp[i - 1]

            if s - target in pref:
                dp[i] = max(dp[i], 1 + (0 if pref[s - target] == -1 else dp[pref[s - target]]))

            pref[s] = i

        return dp[-1]       